### ISSUE TITLE
Write xunit-driven integration tests for shared host tooling

### ISSUE THREAD

**Issue by per-samuelsson** [Starcounter/Starcounter.Nova.Cli#55]
As discussed starting [here](https://github.com/Starcounter/Starcounter.Nova/issues/435#issuecomment-418656635).

I have done some prototyping already, and will do this right after https://github.com/Starcounter/Starcounter.Nova/issues/435 is done, since it depends on that. 

Several sample projects and scripts exist already (in [Nova/test/TestProjects](https://github.com/Starcounter/Starcounter.Nova/tree/develop/test/TestProjects)), so my idea is to turn those into Xunit, drop the scripts, and add a few more.

**Comment by per-samuelsson** [comment]
Process management and diagnostics when any integration test goes wrong has turned out to be quite a pain. On more then one occupation while writing these tests, I've found myself in a state where some processes linger.

I'm also quite rigorous here: I don't want to see these tests on build servers if we can't properly ensure we don't end up in such situations there.

As an example, consider running "myapp,csproj" in a shared host instance. Here's the chain for that (numbers indicate processes).

`dotnet test (1) -> (test code: generate project) -> dotnet run [generated_project.csproj] (2) -> dotnet exec [project.exe] (3) -> scdata.exe (4)`

I considered alternatives, trying to simplify, but haven't really come up with something.

One thing I thought about is if we should bring in [this](https://github.com/dotnet/cli/blob/master/test/Microsoft.DotNet.Tools.Tests.Utilities/Extensions/ProcessExtensions.cs). Then integration test code could at least kill the entire tree if it's not behaving, starting from `dotnet run`.

cc @bigwad @linkdata @miyconst

**Comment by linkdata** [comment]
When considering this, remember that if process (2) dies, for example, then (3) and (4) are [orphaned ](https://en.wikipedia.org/wiki/Orphan_process) and can't be reached through their (now dead) parent process, so the linked-to process-tree-killer would fail.

That's why Nova and level1 test cleanup is based on process names.

**Comment by per-samuelsson** [comment]
> dotnet test (1) -> (test code: generate project) -> dotnet run [generated_project.csproj] (2) -> dotnet exec [project.exe] (3) -> scdata.exe (4)

I came up with one improvement at least, changing the above chain above to:

1. Generate project: sharedhost.csproj
2. Build project: sharedhost.dll
3. Exec project

That will also work better in general, because it mean we can optimize in the future and maybe skip [1+2] if we find inputs are equal. Plus that it's better to report errors / phase (i.e. when building generated project is a step in its own).

**Comment by per-samuelsson** [comment]
> Did it cause any problems?

I did. At some point, all tests failed constantly and I found that it was because some named event could not be properly created. When I finally found what processes to terminate, tests went back to normal.

It might be so that we can't cover all possibilities, yes. But I think it's good at least to gather an understanding of what could go wrong, and how to resolve it.

**Comment by bigwad** [comment]
Do you by any chance share a database between tests?

**Comment by per-samuelsson** [comment]
> Do you by any chance share a database between tests?

Yes, could be so. Not sure how representative that failure is either. So don't pay to much attention.

**Comment by per-samuelsson** [comment]
I think I have enough for a PR soon.

One thing I want to just polish on first is to make sure integration tests are more robust to errors happening, when there are file system stuff that should be dropped (like test database, generated projects and packages). But it shouldn't be to much work.

Then, when I PR, integration tests will only run on an installed "default" database, which is of course unacceptable. But I need #85 to be done in order to be able to pass a custom database name/path to the shared host in tests. If lucky, I can do that right away after this. 😎
